### PR TITLE
cdc: reduce context switch for stateful events

### LINKED ISSUE #5610: cdc goroutines become crazy when initializing a capture with lots of regions and tikv.cdc.min-ts-interval = 200ms
### What did you do?

Create a capture on 16 MySQL tables, 131K regions totally with
* tikv.cdc.min-ts-interval = 200ms
* tikv.cdc.min-ts-interval = 1s

### What did you expect to see?

Both 2 operations can be finished quickly. With `tikv.cdc.min-ts-interval = 1s` it's expected:
图片

### What did you see instead?

However with `tikv.cdc.min-ts-interval = 200ms` the capture creating can never finished, and cpu usage is *very high*:
图片
图片

And here is a 30s CPU profile:
图片

We can see the CPU time is about 480s, which means all 16C are very busy on `runtime.selectgo`.

### Versions of the cluster

All components are nightly.
Go version is 1.18.1.

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Issue Number: close #5610 

#### Tests 

 - No code

- Manual test with the issued case:
  
  图片
  
  图片

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/kv/client.go [hunk 1]
```diff
@@ -1275,76 +1273,97 @@ func (s *eventFeedSession) receiveFromStream(
...
 
-	if isNewSubscription {
-		// It's the first response for this region. If the region is newly connected, the region info should
-		// have been put in `pendingRegions`. So here we load the region info from `pendingRegions` and start
-		// a new goroutine to handle messages from this region.
-		// Firstly load the region info.
-		state, ok = pendingRegions.take(event.RequestId)
-		if !ok {
-			log.Warn("drop event due to region feed is removed",
+		if isNewSubscription {
```
Comment (lonng) [inline cdc/kv/client.go:1315]: Can we use `else` directly instead of two `if` statements?

File: cdc/kv/client.go [hunk 2]
```diff
@@ -1275,76 +1273,97 @@ func (s *eventFeedSession) receiveFromStream(
...
+	buffSize := len(events)/worker.concurrent + 1
+	var statefulEvents []*regionStatefulEvent = nil
+	var buffCap int = 0
+	for i, event := range events {
+		ignore := false
+		state, ok := worker.getRegionState(event.RegionId)
+		// Every region's range is locked before sending requests and unlocked after exiting, and the requestID
+		// is allocated while holding the range lock. Therefore the requestID is always incrementing. If a region
+		// is receiving messages with different requestID, only the messages with the larges requestID is valid.
+		isNewSubscription := !ok
```
Comment (lonng) [inline cdc/kv/client.go:1292]: Why not name the `ok` to `subscripted` to eliminate the redundant variable?

**Comment by asddongmen** [comment]
cdc/kv/region_worker.go:530:28: Error return value of `(github.com/pingcap/tiflow/pkg/workerpool.EventHandle).AddEvent` is not checked (errcheck)

[2022-06-02T07:15:24.648Z] 			w.handles[slot].AddEvent(ctx, &regionStatefulEvent{finishedCallbackCh: finishedCallbackCh})

**Review by asddongmen — CHANGES_REQUESTED** [review]
plz fix lint check, rest LGTM

File: cdc/kv/client.go [hunk 3]
```diff
@@ -1235,10 +1235,9 @@ func (s *eventFeedSession) receiveFromStream(
 			s.deleteStream(addr)
 
 			// send nil regionStatefulEvent to signal worker exit
-			select {
-			case worker.inputCh <- nil:
-			case <-ctx.Done():
-				return ctx.Err()
+			err = worker.retrieveEvents(ctx, []*regionStatefulEvent{nil})
```
Comment (overvenus) [inline cdc/kv/client.go:1238]: The comment and the method name conflicts. "retrieve" does not mean "send".

File: cdc/kv/client.go [hunk 4]
```diff
@@ -1275,76 +1272,87 @@ func (s *eventFeedSession) receiveFromStream(
...
-				zap.String("changefeed", s.client.changefeed.ID),
-				zap.Uint64("regionID", event.RegionId),
-				zap.Uint64("requestID", event.RequestId),
-				zap.Uint64("currRequestID", state.requestID),
-				zap.String("addr", addr))
-			return nil
-		}
+	statefulEvents := make([][]*regionStatefulEvent, worker.inputSlots)
+	for i := 0; i < worker.inputSlots; i++ {
+		buffLen := len(events) / worker.inputSlots * 3 / 2
```
Comment (overvenus) [inline cdc/kv/client.go:1284]: Could you add some comments about the formula?

**Review by asddongmen — APPROVED** [review]

**Review by overvenus — APPROVED** [review]
Rest LGTM

File: cdc/kv/region_worker.go [hunk 5]
```diff
@@ -832,6 +837,19 @@ func (w *regionWorker) evictAllRegions() {
...
+// sendEvents puts events into inputCh and updates some internal states.
+// Callers must ensure that all items in events can be hashed into one handle slot.
+func (w *regionWorker) sendEvents(ctx context.Context, events []*regionStatefulEvent) error {
+	atomic.AddInt32(&w.inputPending, int32(len(events)))
+	select {
+	case <-ctx.Done():
+		atomic.AddInt32(&w.inputPending, -int32(len(events)))
+		return ctx.Err()
+	case w.inputCh <- events:
+		return nil
```
[the remark this answers is not in the document]
Comment (hicqu) [reply cdc/kv/region_worker.go:849]: No. We sub it when fetching items from the channel.

**Pull request merged**